[Merged by Bors] - FreeBSD compatibility

tzdybal opened the pull request:
## Motivation
This PR makes it possible to compile and run `go-spacemesh` on FreeBSD.

## Changes
Some fixes to makes build scripts more portable.
Temporal update `post` of dependency, before spacemeshos/post#38 is merged.

## Test Plan
1. Checkout code on FreeBSD machine.
2. Run `gmake`.
3. Run `./build/go-spacemesh`

I'm running my FreeBSD node for some time, without any issues.

tzdybal committed:
Portable use of 'bash' in build scripts

tzdybal committed:
FreeBSD-compatibile handling of `protoc`

tzdybal committed:
Temporal replacement of `post` dependency

This can be removed after merge of
https://github.com/spacemeshos/post/pull/38

lrettig committed:
typo

lrettig commented on scripts/genproto.sh:
```diff
@@ -27,11 +27,14 @@ if [ ! -f $protoc ] ; then
 
 echo "using protoc from $protoc"
+if [[ $protoc == "protoc" ]]; then
+	devtools_opts="-I./devtools/include/"
```
This directory does not exist, and will never be created on FreeBSD, since scripts/install-protobuf.sh exits whenever `$kernel == FreeBSD`

tzdybal replied:
I had a dirty workspace (from previous tests), that's why it worked locally. It's fixed by 3ee68e7.

lrettig commented on go.mod:
```diff
@@ -42,4 +42,6 @@ require (
 	nanomsg.org/go-mangos v1.4.0
 )
 
+replace github.com/spacemeshos/post => github.com/tzdybal/post v0.0.0-20200629203636-775866cafc8f
```
Let's merge https://github.com/spacemeshos/post/pull/38 first

lrettig reviewed (changes requested):
Thanks for this PR! I'm getting errors from protobuf when trying to compile the proto files on FreeBSD (see also my comment about `devtools/` missing):

```
% ./scripts/genproto.sh
Could not find protoc in ./devtools/bin/protoc. Trying to use protoc in PATH.
using protoc from protoc
Generating protobuf for api/pb
./devtools/include/: warning: directory does not exist.
google/protobuf/descriptor.proto: File not found.
google/api/annotations.proto:20:1: Import "google/protobuf/descriptor.proto" was not found or had errors.
google/api/annotations.proto:28:8: "google.protobuf.MethodOptions" is not defined.
google/protobuf/empty.proto: File not found.
api/pb/api.proto:4:1: Import "google/api/annotations.proto" was not found or had errors.
api/pb/api.proto:5:1: Import "google/protobuf/empty.proto" was not found or had errors.
api/pb/api.proto:180:25: "google.protobuf.Empty" is not defined.
api/pb/api.proto:186:24: "google.protobuf.Empty" is not defined.
api/pb/api.proto:192:25: "google.protobuf.Empty" is not defined.
api/pb/api.proto:198:28: "google.protobuf.Empty" is not defined.
api/pb/api.proto:222:20: "google.protobuf.Empty" is not defined.
api/pb/api.proto:228:23: "google.protobuf.Empty" is not defined.
```

Also it looks like some of the changes are causing the build to fail on Linux - see [this Travis run](https://travis-ci.org/github/spacemeshos/go-spacemesh/jobs/703624891):

```
$ make install
./setup_env.sh
/usr/bin/env: invalid option -- 'S'
Try '/usr/bin/env --help' for more information.
Makefile:40: recipe for target 'install' failed
make: *** [install] Error 125
```

lrettig commented:
Note to self: packages I had to install on FreeBSD to get this far: git, bash, gmake, protobuf, go

tzdybal committed:
genproto.sh: fix FreeBSD include paths

tzdybal commented:
I guess that problem with `-S` is related to `busybox` implementation of `env`.
`env` from `coreutils` handles `-S` correctly.

tzdybal committed:
Even more portable use of 'bash' in build scripts

`/usr/bin/env` is provided by busybox and coreutils. The option '-S' is
available only in coreutils version.

More generic way of setting "exit immediately on error" is to start bash
without arguments and then call `set -e` bash command.

lrettig commented:
Thanks! Tested and confirmed that this is now working on freebsd, linux, and macos. I think the travis failure is due to #2050. I'd like to merge spacemeshos/post#38 before merging this so as not to require the module replace, will take a look at that.

lrettig committed:
Add freebsd to platforms list

lrettig committed:
Add FreeBSD build instructions

lrettig committed:
Upgrade to latest post library

Now that https://github.com/spacemeshos/post/pull/38 has been merged

lrettig commented:
bors try

lrettig commented:
@tzdybal I think this is about ready to merge. I just updated the post module and rebased on the latest develop branch. Could you please give this one final test and make sure it's still working for you on freebsd?

lrettig commented:
bors try

>

lrettig commented:
bors try

lrettig commented:
Tested again on freebsd, it's working for me, will merge

lrettig commented:
bors merge

noamnelke commented on go.mod:
```diff
@@ -22,7 +22,7 @@ require (
 	github.com/spacemeshos/merkle-tree v0.0.0-20191028110812-1908c3126c82
 	github.com/spacemeshos/poet v0.1.0
-	github.com/spacemeshos/post v0.0.0-20191225190235-dfb8a5803e6d
+	github.com/spacemeshos/post v0.0.0-20200707150818-013318bab6f4
```
This seems unrelated...

noamnelke replied:
oh, I just figured out why it is actually related (https://github.com/spacemeshos/post/pull/38)... 👍🏻

noamnelke replied:
Note to self: read existing comments before posting your own 😆
